### PR TITLE
lib/types: add the selectorFunction type

### PR DESCRIPTION
**Pull request by basvandijk**
This came out of https://github.com/NixOS/nixpkgs/pull/38698#issuecomment-410894310.

This is used to type options like: `services.xserver.windowManager.xmonad.extraPackages` that specify functions that take an attribute set containing packages / plugins and return a list containing a selection of the values in this set.

###### Motivation for this change

The reason we need a dedicated type for this is to define the correct merge behaviour. Without the `selectorFunction` type merging multiple selector function option definitions results in an evaluation error.
The `selectorFunction` merges definitions by returning a new selector function that applies the selector functions of all the definitions to the given input and concatenates the results.

###### Bikeshedding

Instead of naming it `selectorFunction` we could name it `selector` which is shorter. However I opted for the former because it makes it immediately clear the value should be a function. 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Infinisil** [comment]
Ping @Profpatsch @rycee 

This looks good to me, neat!

**Comment by oxij** [comment]
Isn't this just a less generic version of `functionTo` that existed before 4ff1ab5a56f1280d2de319ad4eb4b2796e07ed35?

In https://github.com/NixOS/nixpkgs/pull/11866/files#r48413164 @nbp writes on the topic:

> no function should be allowed in the module system!
> Function should be forbidden by default [...] because they are impossible to document properly, and to introspect statically.

I disagree, IMHO function (X → Monoid) is as good a Monoid as any other and hence `types` should provide such an instance. While I agree that overusing `functionTo` in a module system isn't good, that type is pretty useful.

In short, we revived it SLNOS by reverting the above commit. This PR shows that there are uses even in pure NixOS. I think you should just revert 4ff1ab5a56f1280d2de319ad4eb4b2796e07ed35 instead.

**Comment by basvandijk** [comment]
@oxij great suggestion. I created https://github.com/NixOS/nixpkgs/pull/44601 as an alternative.

**Comment by nbp** [comment]
No, we should avoid having function as NixOS option. While this might make our life simpler to think in terms of functions, not everybody has a degree in functional programming, and we should not require as much from our users. If you disagree, you are free to have your own modules outside of NixOS, or to use one of the following suggestions:

For all the uses cases mentioned here, a simpler alternative is to add an overlay to add the set of packages out of which compatible packages can be extracted, such as `pkgs.ihaskellPkgs.*` or `pkgs.xmonadPkgs.*`.  `pkgs` is the attribute set from which users are expected to take packages out of, not any argument given to any option as argument. Arguments of function are not extensible with the module system. On the other hand, `pkgs` packages are modular, to the extend that overlays are ordered properly.

Another option, is to create an attribute set which contains the packages to be used as argument, like we do with `system.build` (https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/activation/top-level.nix#L144). However, such attribute set as the same limitation as function argument as not being extensible.

Sorry, if this answer is not what you had expected, but until I see these alternative fail (i-e not being as expressive or more), I will consider this answer to be final.

**Comment by Profpatsch** [comment]
@oxij is right that this is a special case of (one sensible?) function instance for Monoid, that is “a function which takes some `T` and returns something that is a Monoid and can be merged”:

```
default :: Monoid m => a -> m
default = const default
merge :: Monoid m => (a -> m) -> (a -> m) -> (a -> m)
merge f g x = merge (f x) (g x)
```

In our case `a -> m` is `Attrs a -> List a`, which specializes in two dimensions: 1) the input can only be an attribute set and 2) the output list can only contain the same elements as the input set.
We could think of a lot more generalizations of `a -> m`, `a` can be any type by definition and `m` can be any type that exists in `types.nix`, because they all implement `merge` and `default` and are therefore Monoids by construction.

You can easily achieve that generality in your definition as well, by giving `selectorFunction` an argument type `t` that specifies the output type and then copying how e.g. `coercedTo` defines their `merge` function (combining the arguments with `t.merge` instead of `concatMap`).

**Comment by nbp** [comment]
@Profpatsch Honestly, if we have to explain category theory to understand NixOS, then this project is a failure. I am glad that people can have satisfaction in the sanity of the idea using these. However, this does mean that we should live up to everything which can be expressed in this system, especially if this leads to a documentation hell.

Today, I think that functions are a documentation hell to explain to someone who has only manipulated `apt-get install` commands so far. So, if the alternative are disproved after being tested, then we can re-open this discussion.

**Comment by oxij** [comment]
Let me preemptively remind you that I mean no personal offense. I see the argument, I disagree with the argument, I dump by thoughts about the argument. Such things are much easier to do via another anonymous peer review system, but NixOS uses GitHub which insists on contributors having names with all the attached personality-related anthropological baggage.

Let me also remind you that the service layer implementation of the module system is one of the three main disagreements between the current NixOS and SLNOS (with the other two being use flags and systemd), so I'm obviously biased.

@nbp

> For all the uses cases mentioned here, a simpler alternative is to add an overlay ...

By that same logic, instead of providing `boot.kernelPackages` option we should make users override `pkgs.linuxPackages` with overlays. Such a design reverses the normal relation where nixpkgs provides and NixOS modules use. What I hear you saying is that you want to introduce NixOS-module-specific overlays to nixpkgs for every module that manages an overridable lists of packages. That's way worse than `functionTo`, see below.

> Another option, is to create an attribute set which contains the packages to be used as argument

So, instead of

```nix
whateverPackages = p: with p; [ a b c ];
```

users would have to write

```nix
whateverPackagesSet = pkgs.whateverPackagesSet_version;
whateverPackages = with config.whateverPackagesSet; [ a b c ];
```

Isn't that just an ugly encoding of the same function to monoid by tying-the-knot over `config.whateverPackageSet` with a module system itself?

Yes, it is. And your previous solution does the same by tying-the-knot over `pkgs` argument of the module system with overlays.

I also feel like this encoding even apparently looks more complicated than the original. You had several tiiiiny local functions that produced lists, but to hide the fact that they are functions you now have to encode their arguments with huuuuuge fixed points over the whole module system (or via overlays over the `pkgs` argument of the module system) and set/override some seemingly unrelated things.

Also note that neither solution works for multi-instance services (which NixOS doesn't implement ATM, I know, but see below), in both cases you won't be able to give names to your overlays/system.build-like options beforehand.

> ... but until I see these alternative fail ...

Define "fail". These alternatives fail modularity and KISS-principle checks: they encode the same monoids, but with worse encodings.

To paraphrase Ben Franklin: "Those who would give up Sane Semantics to purchase a little Apparent Simplicity, will have neither good Semantics nor actual Simplicity."

> ... if we have to explain category theory to understand NixOS ...

1) Neither functions nor monoids are "Category theory".

2) To "understand" NixOS module system one would need to understand both of them anyway (one might reason in an equivalent system, but if one understands how module system works one most certainly understands both already).

3) "Understand" != "Use", to use `functionTo` you don't need to understand what it does. Most users won't even be aware that it exists, even when using the options in question, as witnessed by the current state of having said options without anyone complaining (unless they try to assign them from several places and the module system which has no types for them starts to panic).

> I think that functions are a documentation hell to explain to someone who has only manipulated `apt-get install` commands so far

That's not an argument.

1) You don't need to explain, just give examples, newbies copy-paste and modify ready-made examples.

2) Also note that you can't plainly explain how your solutions work either, the user has to set/override some seemingly unrelated things (overlays, some random variables) for them to work properly.

"Hey, so why do I have to use `config.` before `whateverPackageSet` in the assignment to `whateverPackages` to make it work?"

"Because that is not an assignment and `config.` is a dark secret art of the module system you'll never understand before you understand functions, fixed points, and mkMerge attrset monoid anyway. So, please go learn some functional programming first."

Btw, this reminds me that you criticized monoidal services design with a similar "grand-mother" argument (https://nixos.org/nix-dev/2014-December/015370.html). Do you, like, hate monoids or something?

**Comment by basvandijk** [comment]
@nbp regarding your first suggestion:

> a simpler alternative is to add an overlay to add the set of packages out of which compatible packages can be extracted, such as `pkgs.ihaskellPkgs.*` or `pkgs.xmonadPkgs.*`

The problem with this approach is that you make the user responsible for choosing the correct package set which is error prone. For example say a users defines the following configuration:

```nix
{ pkgs, ... }: {
   services.postgresql.package = pkgs.postgresql10;
   services.postgresql.extraPlugins = [ pkgs.postgis ];
}
```

This will crash the PostgreSQL server because the `postgis` extension is compiled against a version of `postgresql` which is different than `postgresql10`. Currently our users are supposed to fix this by overriding the extension with the PostgreSQL package in use:

```nix
{ pkgs, ... }: {
   services.postgresql.package = pkgs.postgresql10;
   services.postgresql.extraPlugins = [ 
    (pkgs.postgis.override { postgresql = pkgs.postgresql10; }) 
  ];
}
```

We shouldn't require our users to go to this trouble. In https://github.com/NixOS/nixpkgs/pull/38698 @thoughtpolice proposes to fix this by having users specify the following:

```nix
{ pkgs, ... }: {
   services.postgresql.packages = pkgs.postgresql10Packages;
   services.postgresql.plugins = ps: [ ps.postgis ];
}
```



The postgresql module will then take care of applying the plugins function to the choosen postgresql package set. This approach is more safe and it will become extensible by typing the option with `selectorFunction` or `functionTo`.

Difficulty argument
----------------------

Regarding your argument that functions are too hard to use for regular NixOS users, aren't we already requiring that users understand functions by having them write configurations like: 

```nix
{ pkgs, ... }: { ... }
```

Also having good documentation helps a lot with overcoming this problem. IMHO @thoughtpolice did a good job with this in the [documentation of `services.postgresql.plugins`](https://github.com/NixOS/nixpkgs/blob/2a3d44aeaca886c2d61187d52a66a094840258ba/nixos/modules/services/databases/postgresql.xml#L103:L141).

Alternative
--------------

A potential alternative which doesn't use functions can look like this:

```nix
{ pkgs, ... }: {
   services.postgresql.packages = pkgs.postgresql10Packages;
   services.postgresql.plugins = [ "postgis" ];
}
```

So instead of using a function that selects a list of attributes we use a list of strings instead. The strings will denote the names of attributes which the module will then select out of the correct package set.

In one way this will be safer since we can't accidentally return an attribute from a set different than the input set. Note that we [make this error in the xmonad test](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/xmonad.nix#L14).

I do wonder what kind of error messages users will get when they misspell an attribute name: 
`[ "postgresqk" ]` and if this is worse than the error we get when using a function: 
`ps: [ ps.postgresqk ]`...

@nbp @thoughtpolice do you see any problems with this approach.

**Comment by nbp** [comment]
@basvandijk I understand the problem. Thanks for explaining.
The alternative is appealing to me, but list are known to not be modular, and I was wondering wether we could use an (free-form) attribute set of booleans to toggle plugins.

```nix
{ pkgs, ... }: {
   services.postgresql.packages = pkgs.postgresql10Packages;
   services.postgresql.plugins.postgis = true;
}
```

Then we can add a module assertion that all of the attributes provided that way are listed in the list of associated packages.

I think that we can provide a better error message than what Nix does by default when looking up attribute sets. Thus, not using `ps.postgris` should help us improve the error messages friendliness.

**Comment by basvandijk** [comment]
@nbp I like your suggestion. An additional advantage of having an explicit option for each plugin is that we can use it to add additional options per plugin. So we could use a scheme like for example:

```nix
{ pkgs, ... }: {
   services.postgresql.packages = pkgs.postgresql10Packages;
   services.postgresql.plugins.postgis = {
     enable = true;
     ... additional postgis options ...
   };
}
```

@thoughtpolice would this work for you?

**Comment by thoughtpolice** [comment]
It probably would, and I think this change can basically be isolated within the module, but there are some things to think about. I guess this is the only way you can make better error messages work (because you can throw an error based on `passthru.versionCheck`), but I somewhat hate bloating up the module system (and manual) with options like this if possible.

I don't think the ability to specify per-plugin options is super useful _right now_, but maybe in the future some other code could make use of it. It might be most useful for overrides (i.e. "I want to apply this patch to my Postgis plugin"), but probably not for most configuration ("I need to set `shared_preload_libraries` in `postgresql.conf`").

**Comment by nbp** [comment]
@thoughtpolice bloating up the module system should not be an issue. What would be an issue is users not being to figure out that NixOS support the feature they want because of the lack of documentation. For example that you can easily create mediawiki instances: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/apache-httpd/mediawiki.nix

> the module system has massive problems, including being "non modular" to begin with; good luck starting two instances of Postgres, for instance.

Do not blame the module system for not being modular by the fact that NixOS does not have the features you are looking for. I bet one can write a `modular-postgres.nix` module and use it in NixOS.

**Comment by danbst** [comment]
I  don't get "serializability" argument. As of now we can't get plain JSON out of configuration
```
$ nix-instantiate --eval -E 'with import <nixpkgs/nixos> { configuration.boot.isContainer = true; }; 
      config' --strict
trace: lib.zip is deprecated, use lib.zipAttrsWith instead
trace: `mkStrict' is obsolete; use `mkOverride 0' instead.
trace: `lib.nixpkgsVersion` is deprecated, use `lib.version` instead!
trace: types.optionSet is deprecated; use types.submodule instead
trace: `types.list` is deprecated; use `types.listOf` instead
trace: Warning: `showVal` is deprecated and will be removed in the next release, please use `traceSeqN`
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
trace: WARNING: `stdenv.isArm` is deprecated after 18.03. Please use `stdenv.isAarch32` instead
error: enum-0.4.7 not supported for interpreter python3.7
(use '--show-trace' to show detailed location information)
```
```
$ nix-instantiate --eval -E 'with import <nixpkgs/nixos> { configuration.boot.isContainer = true; }; 
        config.services' --strict
error: The option `services.networking.websockify.sslCert' is used but not defined.
(use '--show-trace' to show detailed location information)
```
```
$ nix-instantiate --eval -E 'with import <nixpkgs/nixos> { configuration.boot.isContainer = true; }; 
      config.assertions' --strict
error: attribute 'cycle' missing, at /nix/var/nix/profiles/per-user/root/channels/nixpkgs/nixos/modules/tasks/filesystems.nix:214:119
(use '--show-trace' to show detailed location information)
````
We can't serialize derivations, so need `literalExample` hack. So adding a function type wouldn't make things worse.

As for selector function, I'd like to have 2 variants: `types.listSelector` and `types.attrsSelector`. Both represent a function, that takes an attrset input parameter, but produce list and attrs respectively. *Mabe* we can make it parametrized type (`types.listSelector pkgs`), but I'm not sure it is possible. The postgresql example would look then like:
```
options.pg = mkOption { default = pkgs.postgresql; };
options.plugins = mkOption {
  type = types.listSelector config.pg.pkgs;
  literalExample = ''pkgs: [ pkgs.pg_repack ]'';
  default = _: [];
};
config = {
  ... "${config.pg.withPackages config.plugins}/bin/postgres ....
};
```
Actually, I like nixsap style more:
```
options.plugins = mkOption {
  type = types.attrsSelector config.pg.pkgs;
  literalExample = ''pkgs: { inherit (pkgs) pg_repack; }'';
  default = _: {};
};
config = {
  ... "${config.pg.withAttrPackages config.plugins}/bin/postgres ....
};
```

PS: this was already started by @nbp [8 years ago](https://github.com/NixOS/nixpkgs/commit/6669cf61f44d750677462d46ee1ac987b7cae0d6)

https://github.com/NixOS/nixpkgs/blob/db7017362776cd04fc07ae89ec2d00f247df5ecd/nixos/modules/misc/nixpkgs.nix#L17-L43

**Comment by basvandijk** [comment]
@danbst regarding:

>  I'd like to have 2 variants: types.listSelector and types.attrsSelector

We could also reconsider https://github.com/NixOS/nixpkgs/pull/44626 which provides `functionTo` which is a generalisation of `selectorFunction`. With that we can have:

* `listSelector = xs: functionTo (listOf xs)` 
* `attrsSelector = attrs: functionTo (attrsOf attrs)`

**Comment by danbst** [comment]
@basvandijk that's fine. What I've proposed is allow dependent types (muahaha) in module system. Then, if we do something like this:
```
      plugins = mkOption {
        type = listSelector cfg.package.pkgs;
        description = "bla";
      };
```
we can infer argument type from `cfg.package.pkgs`. I recall @ts468 implemented such infer function before.

Added benefit is that we can introspect that set, for example, in future interactive configurator:
```
$ nix-instantiate --eval ./nixos --arg configuration '{
     boot.isContainer = true; 
     services.postgresql.enable = true;
     services.postgresql.plugins = ps: with ps; [ pg_repack ];

}' -A options.services.postgresql.plugins.type.value -A options.services.postgresql.plugins.type.arg1

<LAMBDA>
{ cstore_fdw = <CODE>; pg_cron = <CODE>; pg_hll = <CODE>; pg_repack = <CODE>; pg_similarity = <CODE>; pg_topn = <CODE>; pgjwt = <CODE>; pgroonga = <CODE>; pgtap = <CODE>; plv8 = <CODE>; postgis = <CODE>; timescaledb = <CODE>; tsearch_extras = <CODE>; }
```

**Comment by turion** [comment]
I recently implemented something like this ad hoc in https://github.com/NixOS/nixpkgs/pull/109035. I believe it should be in `lib/types.nix`. Is there any progress on this?

**Comment by Profpatsch** [comment]
Pinging @Infinisil, since he has the best overview of the types setup right now.

**Comment by Infinisil** [comment]
See https://github.com/NixOS/nixpkgs/pull/110707

**Comment by Infinisil** [comment]
This has been implemented with https://github.com/NixOS/nixpkgs/pull/110707!
